### PR TITLE
Improved Client handling of failed RPCs

### PR DESCRIPTION
**Pull request by dadgar**
This PR improves the handling of failed RPCs by:
1. Randomizing servers such that we don't repick the same server if they are
being set from Consul.
2. Trigger a retry when there is a sucessful RPC.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -137,9 +137,11 @@ type Client struct {
...
 	triggerEmitNodeEvent chan *structs.NodeEvent
 
-	// discovered will be ticked whenever Consul discovery completes
-	// successfully
-	serversDiscoveredCh chan struct{}
+	// rpcRetryCh is closed when there an event such as server discovery or a
+	// successful RPC occurring happens such that a retry should happen. Access
+	// should only occur via the getter method
+	rpcRetryCh   chan struct{}
+	rpcRetryLock sync.Mutex
```
Comment (schmichael) [inline client/client.go]: I don't think closing+niling is functionally any different from ticking, and it's more complex.

File: client/rpc.go [hunk 2]
```diff
@@ -67,6 +67,7 @@ TRY:
 	// Make the request.
 	rpcErr := c.connPool.RPC(c.Region(), server.Addr, c.RPCMajorVersion(), method, args, reply)
 	if rpcErr == nil {
+		c.fireRpcRetryWatcher()
```
Comment (schmichael) [inline client/rpc.go]: Why are you firing off a retry here when there wasn't an error?

File: client/servers/manager.go [hunk 3]
```diff
@@ -74,6 +74,16 @@ func (s *Server) String() string {
 	return s.addr
 }
 
+func (s *Server) Equal(o *Server) bool {
+	if s == nil && o == nil {
+		return true
+	} else if s == nil && o != nil || s != nil && o == nil {
+		return false
+	}
```
Comment (schmichael) [inline client/servers/manager.go]: Can be shortened to `if s == nil || o == nil { return s == o }`

h/t to @chelseakomlo

File: client/servers/manager.go [hunk 4]
```diff
@@ -157,10 +194,24 @@ func (m *Manager) Start() {
...
+
+	// Sort both the  existing and incoming servers
+	servers.Sort()
+	m.servers.Sort()
+
+	// Determine if they are equal
+	equal := servers.Equal(m.servers)
+
+	// Randomize the incoming servers
+	servers.shuffle()
```
Comment (schmichael) [inline client/servers/manager.go]: Do you still want to shuffle if they're equal?

Comment (dadgar) [reply client/servers/manager.go]: Yeah otherwise we will have deterministic ordering through Consul discovery.

**Pull request merged**